hphp/runtime/vm/jit/irgen-builtin.cpp

```diff
@@ -488,6 +473,113 @@ SSATmp* opt_set_frame_metadata(IRGS& env, uint32_t numArgs) {
   return cns(env, TInitNull);
 }
 
+SSATmp* opt_foldable(IRGS& env,
+                     const Func* func,
+                     uint32_t numArgs,
+                     uint32_t numNonDefaultArgs) {
+  auto const constAsCell = [] (const SSATmp* tmp) {
+    switch (tmp->type().toDataType()) {
+      case KindOfUninit:
+        return make_tv<KindOfUninit>();
+      case KindOfNull:
+        return make_tv<KindOfNull>();
+      case KindOfBoolean:
+        return make_tv<KindOfBoolean>(tmp->boolVal());
+      case KindOfInt64:
+        return make_tv<KindOfInt64>(tmp->intVal());
+      case KindOfDouble:
+        return make_tv<KindOfDouble>(tmp->dblVal());
+      case KindOfPersistentString:
+        return make_tv<KindOfPersistentString>(tmp->strVal());
+      case KindOfPersistentArray:
+        return make_tv<KindOfPersistentArray>(tmp->arrVal());
+      case KindOfPersistentVec:
+        return make_tv<KindOfPersistentVec>(tmp->vecVal());
+      case KindOfPersistentDict:
+        return make_tv<KindOfPersistentDict>(tmp->dictVal());
+      case KindOfPersistentKeyset:
+        return make_tv<KindOfPersistentKeyset>(tmp->keysetVal());
+
+      case KindOfVec:
+      case KindOfDict:
+      case KindOfKeyset:
+      case KindOfClass:
+      case KindOfString:
+      case KindOfArray:
+      case KindOfObject:
+      case KindOfResource:
+      case KindOfRef:
+        break;
+    }
+    // Other Kinds are not expected to be ConstVal
+    not_reached();
+  };
+
+  if (!func->isFoldable()) return nullptr;
+
+  // Don't pop the args yet---if the builtin throws at compile time (because
+  // it would raise a warning or something at runtime) we're going to leave
+  // the call alone.
+  PackedArrayInit args(numArgs);
+  for (auto bcOff = BCSPRelOffset{static_cast<int32_t>(numArgs - 1)};
+       bcOff > 
+         BCSPRelOffset{static_cast<int32_t>(numArgs - numNonDefaultArgs - 1)};
+       bcOff--) {
+    auto const t = topType(env, bcOff);
+    if (!t.hasConstVal() && !t.subtypeOfAny(TUninit, TInitNull, TNullptr)) {
+      return nullptr;
+    } else {
+      args.append(cellAsCVarRef(constAsCell(topC(env, bcOff))));
+    }
+  }
+  if (numArgs != func->numNonVariadicParams()) {
+    if (!topType(env).hasConstVal()) return nullptr;
```

Comment:
You should check hasConstVal(TArr), or (since I think the emitter should guarantee that) add an assert that constAsCell(topC(env)) is an array.